### PR TITLE
48 JWT Authentication

### PR DESCRIPTION
**Pull request by danylo-dudok**

### CODE REVIEW

**Review by markiyanshvets — APPROVED** [review]

**Review by markiyanshvets — APPROVED** [review]

**Review by markiyanshvets — APPROVED** [review]

**Review by markiyanshvets — APPROVED** [review]

File: TransIT.BLL/Services/AuthenticationService.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            catch (DbUpdateException e)
+            {
+                _logger.LogError(e, nameof(TokenAsync));
+                return null;
+            }
+            catch (SecurityTokenException e)
+            {
+                _logger.LogError(e, nameof(TokenAsync));
+                return null;
+            }
```
Comment (CrazyRoka) [inline TransIT.BLL/Services/AuthenticationService.cs:101]: Same code
Use `when` keyword

File: TransIT.BLL/Services/AuthenticationService.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                        Create = user
+                    });
+                    await _unitOfWork.SaveAsync();
+                    return token;
+                }
+                return null;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, nameof(TokenAsync));
```
Comment (CrazyRoka) [inline TransIT.BLL/Services/AuthenticationService.cs]: SignInAsync

File: TransIT.BLL/Helpers/JwtFactory.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                var jwtSecurityToken = securityToken as JwtSecurityToken;
+                if (jwtSecurityToken == null
+                    || !jwtSecurityToken.Header.Alg.Equals(
+                        SecurityAlgorithms.HmacSha256,
+                        StringComparison.InvariantCultureIgnoreCase))
+                    throw new SecurityTokenException("Invalid token");
+
+                
+                return (principal, jwtSecurityToken);
+            });
```
Comment (CrazyRoka) [inline TransIT.BLL/Helpers/JwtFactory.cs]: Remove task

File: TransIT.BLL/Helpers/JwtFactory.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        private Task<string> GenerateAccessTokenAsync(int userId, string email, string role) =>
+            Task.Run(() => new JwtSecurityTokenHandler()
+                .WriteToken(new JwtSecurityToken(
+                    issuer: _jwtOptions.Issuer,
+                    audience: _jwtOptions.Audience,
+                    notBefore: DateTime.UtcNow,
+                    claims: GenerateClaims(userId, email, role),
+                    expires: DateTime.UtcNow.Add(TimeSpan.FromMinutes(_jwtOptions.AccessExpirationMins)),
+                    signingCredentials: _jwtOptions.SigningCredentials
+            )));
```
Comment (CrazyRoka) [inline TransIT.BLL/Helpers/JwtFactory.cs]: Remove task

File: TransIT.BLL/Helpers/JwtFactory.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        private Task<string> GenerateRefreshTokenAsync(int userId, string email, string role) =>
+            Task.Run(() => new JwtSecurityTokenHandler()
+                .WriteToken(new JwtSecurityToken(
+                    issuer: _jwtOptions.Issuer,
+                    audience: _jwtOptions.Audience,
+                    notBefore: DateTime.UtcNow,
+                    claims: GenerateClaims(userId, email, role),
+                    expires: DateTime.UtcNow.Add(TimeSpan.FromMinutes(_jwtOptions.RefreshExpirationMins)),
+                    signingCredentials: _jwtOptions.SigningCredentials
+            )));
```
Comment (CrazyRoka) [inline TransIT.BLL/Helpers/JwtFactory.cs]: Remove task

File: TransIT.BLL/Helpers/JwtFactory.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        
+        public Task<TokenDTO> GenerateTokenAsync(int userId, string email, string role) =>
+            Task.Run(async () => new TokenDTO
+                {
+                    AccessToken = await GenerateAccessTokenAsync(userId, email, role),
+                    RefreshToken = await GenerateRefreshTokenAsync(userId, email, role)
+                });
```
Comment (CrazyRoka) [inline TransIT.BLL/Helpers/JwtFactory.cs]: Remove task run

File: TransIT.API/Extensions/AuthenticationExtension.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        private static string GenerateKey(int length)
+        {
+            var characterSet =
+                "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxy0123456789".ToCharArray();
+            if (length < 0)
+                throw new ArgumentException("length must not be negative", "length");
+            if (length > int.MaxValue / 8)
+                throw new ArgumentException("length is too big", "length");
+            if (characterSet == null)
+                throw new ArgumentNullException("characterSet");
```
Comment (CrazyRoka) [inline TransIT.API/Extensions/AuthenticationExtension.cs]: Remove it

File: TransIT.API/Extensions/AuthenticationExtension.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            var characterArray = characterSet.Distinct().ToArray();
+            if (characterArray.Length == 0)
+                throw new ArgumentException("characterSet must not be empty", "characterSet");
```
Comment (CrazyRoka) [inline TransIT.API/Extensions/AuthenticationExtension.cs]: And this too

File: TransIT.API/Controllers/TestController.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    [EnableCors("CorsPolicy")]
+    [Produces("application/json")]
+    [Route("api/[controller]/[action]")]
+    [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme, Roles = "ADMIN")]
+    public class TestController : Controller
+    {
+        [HttpGet]
+        public IActionResult Greeting()
+        {
+            return new ObjectResult("hello");
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/TestController.cs]: Do we need this?

File: TransIT.API/Controllers/AuthenticationController.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            _authenticationService = authenticationService;
+        
+        [HttpPost]
+        public async Task<IActionResult> SignIn([FromBody] LoginViewModel credentials)
+        {
+            if (ModelState.IsValid)
+            {
+                var res = await _authenticationService.SignInAsync(credentials);
+                return res == null
+                    ? Unauthorized()        as IActionResult
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/AuthenticationController.cs]: Bad Request

File: TransIT.API/Controllers/AuthenticationController.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+                    : new ObjectResult(res) as IActionResult;
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/AuthenticationController.cs]: Json()

File: TransIT.API/Controllers/AuthenticationController.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            return BadRequest();
+        }
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/AuthenticationController.cs:32]:
```
if(isvalid){
var res - ...;
if(rest != null) return json();
}
return badRequest();
```

File: TransIT.API/Controllers/AuthenticationController.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+        [HttpPost]
+        public async Task<IActionResult> Token([FromBody] TokenDTO token)
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/AuthenticationController.cs]: RefreshToken

File: TransIT.API/Controllers/AuthenticationController.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            if (ModelState.IsValid)
+            {
+                var res = await _authenticationService.TokenAsync(token);
+                return res == null
+                    ? Unauthorized()        as IActionResult
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/AuthenticationController.cs]: Forbidden

**Review by CrazyRoka — CHANGES_REQUESTED** [review]

**Review by CrazyRoka — APPROVED** [review]

**Pull request merged**